File: src/Logger.ts
import chalk from "chalk";
import axios from "axios";

/**
 * Log Levels
 * 
 * DEBUG: Debugging/verbose messages that should only be logged in development  
 * INFO: Informational messages that should be logged in production  
 * WARN: Warning messages  
 * ERROR: Error messages  
 */
export enum LogLevel {
    DEBUG,
    INFO,
    WARN,
    ERROR
}

/**
 * Log Styles
 * 
 * Typically an appropriate style is chosen based on the log level so you don't need to specify this
 */
export enum LogStyle {
    NORMAL = "[-] ",
    CHECKMARK = "[✔] ",
    PROGRESS = "[>] ",
    HANG = "[~] ",
    WARNING = "[!] ",
}

/**
 * Log a message to the console
 * @param message The message to log
 * @param level The log level. Defaults to LogLevel.INFO
 * @param style The log style. Defaults to an appropriate style based on the log level
 */
export function log(message: string, level: LogLevel = LogLevel.INFO, style?: LogStyle): void {
    if (!style) {
        switch (level) {
            case LogLevel.ERROR:
            case LogLevel.WARN:
                style = LogStyle.WARNING;
                break;
            default:
                style = LogStyle.NORMAL;
                break;
        }
    }

    switch (level) {
        case LogLevel.DEBUG:
            if (process.env.NODE_ENV !== "production") console.log(style + message);
            break;
        case LogLevel.INFO:
            console.log(style + message);
            break;
        case LogLevel.WARN:
            console.warn(chalk.yellowBright(style + message));
            break;
        case LogLevel.ERROR:
            console.error(chalk.redBright(style + message));
            break;
    }
}

export async function logToDiscord(message: string): Promise<void> {
    if (!process.env.DISCORD_WEBHOOK) throw new Error("No Discord webhook URL provided!");

    await axios.post(process.env.DISCORD_WEBHOOK, {
        content: message
    });

    log("Logged message to Discord", LogLevel.DEBUG);
}

File: ketchupbot-updater/Logger.cs
using System.Drawing;
using Console = Colorful.Console;

namespace ketchupbot_updater;

/// <summary>
/// Log levels for logging messages
/// </summary>
public enum LogLevel
{
    Info,
    Warn,
    Error
}

/// <summary>
/// Log styles for logging messages
/// </summary>
public enum LogStyle
{
    Normal,
    Checkmark,
    Progress,
    Hang,
    Warning
}

/// <summary>
/// Logger for logging messages
/// </summary>
public static class Logger
{
    /// <summary>
    /// Log a message to the console using the specified log level and style
    /// </summary>
    /// <remarks>You should surround the call with a #if DEBUG statement to check if it is meant for debugging.</remarks>
    /// <param name="message">The message to print to the console</param>
    /// <param name="level">The log level to use</param>
    /// <param name="style">The log style to use</param>
    /// <exception cref="ArgumentOutOfRangeException"></exception>
    public static void Log(string message, LogLevel level = LogLevel.Info, LogStyle? style = null)
    {
        // Set default style based on log level
        style ??= level switch
        {
            LogLevel.Error or LogLevel.Warn => LogStyle.Warning,
            _ => LogStyle.Normal
        };

        // Set style string based on style
        string styleString = style switch
        {
            LogStyle.Checkmark => @"[✔] ",
            LogStyle.Progress => "[~] ",
            LogStyle.Hang => "[...] ",
            LogStyle.Warning => "[!] ",
            _ => "[-] "
        };

        string finalMessage = styleString + message;

        switch (level)
        {
            case LogLevel.Info:
                Console.WriteLine(finalMessage);
                break;

            case LogLevel.Warn:
                Console.WriteLine(finalMessage, Color.Yellow);
                break;

            case LogLevel.Error:
                Console.Error.WriteLine(finalMessage);
                break;

            default:
                throw new ArgumentOutOfRangeException(nameof(level), level, null);
        }
    }

    public static void LogToDiscord(string message)
    {
        throw new NotImplementedException();
    }

}